bspump/main.py

```diff
@@ -1,98 +1,123 @@
+                # for line in code.split("\n"):
+                for line in list(filter(None, code.split("\n"))):
+                    if line.startswith("!"):
+                        continue
+                    clean_code += line.replace("\t", "    ") + "\n"
+                if not clean_code:
+                    return
+                    # if bspump.jupyter.bitswan_auto_pipeline.get("sink") is not None:
+                if not cls._in_autopipeline:
+                    fout.write(cell_str(clean_code))
                 else:
-                    exec(clean_code, ctx)
-            except Exception:
-                print(f"Error in cell: {cell_number}\n{clean_code}")
-                # print traceback
-                import traceback
-
-                traceback.print_exc()
+                    cls._cell_processor_contents[cls._cell_number] = cell_str(
+                        clean_code, True
+                    )
+                if not cls._in_autopipeline and contains_function_call(
+                    ast.parse(clean_code), "auto_pipeline"
+                ):
+                    cls._in_autopipeline = True
+
+    @classmethod
+    def parse_notebook(cls, ntb, out_path="tmp.py"):
+        cls._cell_number = 0
+        cls._in_autopipeline = False
+        cls._cell_processor_contents = {}
+        with open(out_path, "w") as f:
+            for cell in ntb["cells"]:
+                cls._cell_number += 1
+                cls.parse_cell(cell, f)
+            # for cell_number, pcell in cls._cell_processor_contents:
+            # j = ''.join(list(cls._cell_processor_contents.values()))
+
+            step_func_code = f"""@async_step
+async def processor_internal(inject, event):
+{''.join(list(cls._cell_processor_contents.values()))}    await inject(event)
+"""
+            f.write(step_func_code)
 
 
 def main():
     app = App()  # noqa: F405
     if app.Test:
         bspump.jupyter.bitswan_test_mode.append(True)
 
-    if os.path.exists(app.Notebook):
-        with open(app.Notebook) as nb:
-            notebook = json.load(nb)
-            cell_number = 0
-            for cell in notebook["cells"]:
-                cell_number += 1
-                exec_cell(cell, cell_number, globals())
-    else:
-        print(f"Notebook {app.Notebook} not found")
-
-    if bspump.jupyter.bitswan_auto_pipeline.get("sink") is not None:
-        register_sink(bspump.jupyter.bitswan_auto_pipeline.get("sink"))  # noqa: F405
-        end_pipeline()  # noqa: F405
-
-    app.init_componets()
-    app.run()
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        if os.path.exists(app.Notebook):
+            with open(app.Notebook) as nb:
+                notebook = json.load(nb)
+                NotebookParser.parse_notebook(
+                    notebook, out_path=f"{tmpdirname}/autopipeline_tmp.py"
+                )
+                sys.path.insert(0, tmpdirname)
+                tmp_module = __import__("autopipeline_tmp")  # noqa: F841
```

Comment:
it would be nice if we could somehow convince python that the file is named `main.ipynb` in the stack traces ;)